proof/refine/TcbFlags_R.thy

```diff
@@ -0,0 +1,96 @@
+(*
+ * Copyright 2025, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+(* Lemmas connecting how the design and abstract specifications represent TCB flags.
+   That is, bitwise with a word in the design spec and as a set in the abstract spec.
+*)
+
+theory TcbFlags_R
+imports BaseRefine.Include Move_R
+begin
+
+lemma tcbFlagToWord_bit:
+  "\<exists>n<word_bits. tcbFlagToWord flag = bit n"
+  by (auto simp: tcbFlagToWord_def ex_nat_less_eq word_bits_conv split: tcb_flag.splits simp del: bit_def)
+
+lemma ex_tcbFlagToWord:
+  "tcbFlagMask !! n \<Longrightarrow> \<exists>flag. tcbFlagToWord flag = bit n"
+  by (auto simp: tcbFlagToWord_def tcbFlagMask_def split: tcb_flag.splits if_splits)
+
+lemma tcbFlagToWord_tcbFlagMask:
+  "flag \<in> word_to_tcb_flags tcbFlagMask \<Longrightarrow> tcbFlagToWord flag && tcbFlagMask = tcbFlagToWord flag"
+  by (cut_tac flag=flag in tcbFlagToWord_bit) (fastforce simp: word_to_tcb_flags_def word_eq_iff)
+
+
+lemma word_to_tcb_flags_not:
+  "word_to_tcb_flags (~~ flags) = word_to_tcb_flags tcbFlagMask - word_to_tcb_flags flags"
+  apply (clarsimp simp: word_to_tcb_flags_def intro!: set_eqI)
+  apply (cut_tac flag=x in tcbFlagToWord_bit)
+  apply (fastforce simp: word_eq_iff)
+  done
+
+lemma word_to_tcb_flags_or:
+  "word_to_tcb_flags (flags || flags') = word_to_tcb_flags flags \<union> word_to_tcb_flags flags'"
+  apply (clarsimp simp: word_to_tcb_flags_def intro!: set_eqI)
+  apply (cut_tac flag=x in tcbFlagToWord_bit)
+  apply (fastforce simp: word_eq_iff)
+  done
+
+lemma word_to_tcb_flags_and:
+  "word_to_tcb_flags (flags && flags') = word_to_tcb_flags flags \<inter> word_to_tcb_flags flags'"
+  apply (clarsimp simp: word_to_tcb_flags_def intro!: set_eqI)
+  apply (cut_tac flag=x in tcbFlagToWord_bit)
+  apply (fastforce simp: word_eq_iff)
+  done
+
+lemmas word_to_tcb_flags_simps = word_to_tcb_flags_not word_to_tcb_flags_or word_to_tcb_flags_and
+
+lemma word_to_tcb_flags_and_not:
+  "flags && tcbFlagMask = flags
+   \<Longrightarrow> word_to_tcb_flags (flags && ~~ flags') = word_to_tcb_flags flags - word_to_tcb_flags flags'"
+  apply (clarsimp simp: word_to_tcb_flags_def intro!: set_eqI)
+  apply (cut_tac flag=x in tcbFlagToWord_bit)
+  apply (fastforce simp: word_eq_iff)
+  done
+
+lemma word_to_tcb_flags_tcbFlagMask[simp]:
+  "word_to_tcb_flags flags \<inter> word_to_tcb_flags tcbFlagMask = word_to_tcb_flags flags"
+  apply (clarsimp simp: word_to_tcb_flags_def intro!: set_eqI)
+  apply (cut_tac flag=x in tcbFlagToWord_bit)
+  apply (fastforce simp: word_eq_iff)
+  done
+
+lemma not_nth_is_and_neq_0:
+  "(\<not> (x::'a::len word) !! n) = (x AND 2 ^ n = 0)"
```

Comment:
why `neq_0`? surely it's `eq_0`?
also: WordLib candidate?

Reply:
Oh yeah, this lemma. It's a tweaked version of an existing lemma from `Word_Lib/More_Word.thy`
```
lemma nth_is_and_neq_0:
  "bit (x::'a::len word) n = (x AND 2 ^ n ≠ 0)"
  ```
Looking at it again, you're right that my version should be `eq_0`. What I'm less confident in is if it is generally useful or if I just needed it because I was giving up on fighting with these word proofs. Happy to move it if we do think that it's useful.

Reply:
It can probably go next to `nth_is_and_neq_0` then? Doesn't have to be in this PR, I think at this point I'm fine with a `(* FIXME move after nth_is_and_neq_0 *)` unless you feel like a lib update